plugins/image-prioritizer/helper.php

```diff
@@ -77,6 +77,65 @@ function image_prioritizer_register_tag_visitors( OD_Tag_Visitor_Registry $regis
 	$registry->register( 'image-prioritizer/video', $video_visitor );
 }
 
+/**
+ * Filters the list of Optimization Detective extension module URLs to include the extension for Image Prioritizer.
+ *
+ * @since n.e.x.t
+ *
+ * @param string[]|mixed $extension_module_urls Extension module URLs.
+ * @return string[] Extension module URLs.
+ */
+function image_prioritizer_filter_extension_module_urls( $extension_module_urls ): array {
+	if ( ! is_array( $extension_module_urls ) ) {
+		$extension_module_urls = array();
+	}
+	$extension_module_urls[] = add_query_arg( 'ver', IMAGE_PRIORITIZER_VERSION, plugin_dir_url( __FILE__ ) . image_prioritizer_get_asset_path( 'detect.js' ) );
+	return $extension_module_urls;
+}
+
+/**
+ * Filters additional properties for the element item schema for Optimization Detective.
+ *
+ * @since n.e.x.t
+ *
+ * @param array<string, array{type: string}> $additional_properties Additional properties.
+ * @return array<string, array{type: string}> Additional properties.
+ */
+function image_prioritizer_add_element_item_schema_properties( array $additional_properties ): array {
+	$additional_properties['lcpElementExternalBackgroundImage'] = array(
```

Comment:
Maybe I should have asked this before... But why are we introducing this in Image Prioritizer, when Optimization Detective already has awareness of the LCP element?

It feels like we're duplicating a bit of what Optimization Detective already handles (e.g. in the `detect.js` script the `onLCP` etc.). Why not enhance this via Optimization Detective? Or provide a more targeted extension point for the LCP element specifically that Image Prioritizer could leverage?

Reply:
Humm. Well, the normal case for how data is collected in Optimization Detective is for a tag visitor to mark tags of interest (which result in `data-od-xpath` attributes being emitted on them on the frontend), and then for OD's `detect.js` to collect data about those specific elements to then send back for storage with the URL Metric.

What is needed here is different because we don't know what the element of interest will be, since any element could be LCP with a background image applied with an external stylesheet.

So that's why this new `lcpElementExternalBackgroundImage` root property is being introduced in the URL Metric schema. There is no XPath because the element is not explicitly visited by tag visitors, so it doesn't necessarily appear in `elements`. Now, if it just so happens that there is a tag which shows up in `elements` (which was processed by a tag visitor) which is also the LCP element... it would make sense to store the LCP metric's `url` there as well. However, this would be coincidental and not something we can rely on specifically for optimizing non-inline background images in LCP elements.

In regards to `onLCP`, we could potentially expose that `onLCP` callback (in addition to `onCLS`, `onFCP`, `onINP`, and `onTTFB`) to the `initialize` function for each extension. This would eliminate the need to pass the `webVitalsLibrarySrc`.

Reply:
I may be missing some understanding here, so apologies in advance. But don't we only care about the background image that's relevant to the LCP element?

So from that perspective, couldn't we have an extension point that allows extensions to expand what data is collected for the LCP element? And then Image Prioritizer could add the background image data as needed?

Reply:
The problem is that the tag visitors don't know what the LCP element with an external background image could possibly be, so they don't know to `return true` to flag the tag for capturing in URL Metrics. I mean, there could be a tag visitor that just returns `true` for every visited tag, which would result in `data-od-xpath` attributes being added to every single tag and then for every single tag to be stored in the `elements` of URL Metrics. But this would be wasteful.

In the case of an `IMG` being the LCP element, this is covered because the tag visitors for images are visiting every `IMG` so they are all being captured in the `elements` of a URL Metric. When such an `IMG` is the LCP element, we don't even need the `url` from the LCP Metric because the URL is right there in the `src` and `srcset`. The same goes for an element which has an inline `background-image` style: the tag visitor can see it so optimization can preload what is defined right there in the HTML. 

But for elements with external background images, this is a special case.

Reply:
I've exposed the Web Vitals functions to extensions in 47d06dc669b34ef2cffaef01ad8be2d11cbe06ca.

Reply:
https://github.com/WordPress/performance/commit/47d06dc669b34ef2cffaef01ad8be2d11cbe06ca#diff-ec3ffd3a54ce7df9d216b18a1dee27e4c41785e16698bf3229efcef5e7072fb2R64 looks good to me. It's definitely a nice little enhancement to not have to reimplement that part.

I guess I had not considered this scenario before - so is it correct that, in case the LCP element is not one of the specific elements that are being visited, it will be unknown what the LCP element is for that page load?

In that case, it makes sense that the extension has to implement it. Though I wonder whether we could make this a bit more ergonomic and centralize the implementation, e.g. to get benefits like "common LCP element" out of the box (with what #1723 improved). Doesn't have to be in this PR, but maybe we could implement the LCP element logic centrally so that every URL metric receives the LCP element, whichever it is? If we're worried about overhead of doing this when it's not needed, we could make it opt-in so that it's only recorded if at least one extension opts in to requiring it.

It is a bit specific to implement this directly in OD, but on the other hand my concern is that not doing so will have any extension that needs something LCP related to reinvent the wheel, when instead we could have a general LCP element handler and extensions could only extend what data is included as part of it. WDYT?

Reply:
> I guess I had not considered this scenario before - so is it correct that, in case the LCP element is not one of the specific elements that are being visited, it will be unknown what the LCP element is for that page load?

Yes, that's exactly right. It won't be captured which element is the LCP element since the LCP element is something other than one of the elements being tracked.

> In that case, it makes sense that the extension has to implement it. Though I wonder whether we could make this a bit more ergonomic and centralize the implementation, e.g. to get benefits like "common LCP element" out of the box (with what #1723 improved). Doesn't have to be in this PR, but maybe we could implement the LCP element logic centrally so that every URL metric receives the LCP element, whichever it is? If we're worried about overhead of doing this when it's not needed, we could make it opt-in so that it's only recorded if at least one extension opts in to requiring it.

So you're saying to implement a new root-level property in the URL Metric, like `lcpElement` (alongside the existing `elements`) which captures the properties in the [`LargestContentfulPaint`](https://developer.mozilla.org/en-US/docs/Web/API/LargestContentfulPaint) interface? In this PR that's what is essentially what is being done, although specifically for a background image. And since the `element` can't be captured as-is, this PR is capturing the `id` (also exposed on `LargestContentfulPaint`), `tag`, and `class` as a way to identify the element. Ideally we'd be able to store the server-side XPath to the element, but again it may not be a visited tag, so that `xpath` property could perhaps be a nullable field so that client-side detection can provide it if it so happens that the LCP element has a `data-od-xpath` attribute. In this way, we wouldn't technically need to store `isLCP` with each element, as we could then just look to see if the `xpath` for an element matches the XPath stored with the LCP data.

And if we're capturing these properties of `LargestContentfulPaint`, should we also then capture the other properties like `size`, `renderTime`, and `loadTime`?

And then if we're capturing the LCP metrics, should we then also centralize the capturing of the CLS and INP data as well? 

Or maybe we reconsider for later since this could go to far. For storing the LCP `url`, there's also the question of data validation. We'd want to make sure that we only preload safe URLs which is being addressed in #1713. We can always update Image Prioritizer later to reuse LCP metric data stored by Optimization Detective centrally, especially when/if it comes to light that more use cases need access to this centralized data.

So how about going with an Image Prioritizer-specific implementation but then we open a new issue to discuss capturing these metrics centrally in a future version. This would tie in nicely with the Performance dashboard (#1324) since the data would already be there (although the dashboard would probably want to introduce additional data retention to capture more than the sample size).

Reply:
> So you're saying to implement a new root-level property in the URL Metric, like `lcpElement` (alongside the existing `elements`) which captures the properties in the [`LargestContentfulPaint`](https://developer.mozilla.org/en-US/docs/Web/API/LargestContentfulPaint) interface?

Yes, exactly.

> And if we're capturing these properties of `LargestContentfulPaint`, should we also then capture the other properties like `size`, `renderTime`, and `loadTime`?

Maybe. I think we should probably keep it light in the default implementation, but allow extensions to expand what is detected. For example there should be an extension point where the LCP data is reported in JS so that a plugin could add more data to the `lcpElement` object that is sent to the REST API, and on the server-side it would need to be possible to extend the schema for what can be in `lcpElement`.

I also would prefer not to include actual performance numbers, because OD is primarily an optimization tool, not a monitoring tool. A separate extension for monitoring could add such data as needed.

> And then if we're capturing the LCP metrics, should we then also centralize the capturing of the CLS and INP data as well?

Potentially, but only if we see a need for it. While it seems to make sense from a metric perspective, I'd argue LCP is different from CLS and INP because there is no "CLS element" or "INP element". The LCP element is directly relevant to things that Optimization Detective allows to do. This also relates to what I mention above about focusing on areas that help optimization, not performance metrics. I find it less applicable for INP and CLS, because what would we send there to help optimization? Possibly there is some useful data (e.g. long tasks), but I feel that this should be separately explored, if at all.

> So how about going with an Image Prioritizer-specific implementation but then we open a new issue to discuss capturing these metrics centrally in a future version. This would tie in nicely with the Performance dashboard (#1324) since the data would already be there (although the dashboard would probably want to introduce additional data retention to capture more than the sample size).

That sounds good to me.

Reply:
> there is no "CLS element"

Well, there are elements that cause CLS. So in theory a top-level CLS property could capture the [`LayoutShiftAttribution`](https://developer.mozilla.org/en-US/docs/Web/API/LayoutShiftAttribution) [sources](https://developer.mozilla.org/en-US/docs/Web/API/LayoutShift/sources). Also, for INP  That said, Embed Optimizer isn't using this and is instead using `ResizeObserver`. Things to explore in the new issue. I've just filed it as #1730.